pyproject.toml

```diff
@@ -0,0 +1,66 @@
+[build-system]
+requires = ["poetry-core>=1.1.0b3"]
+build-backend = "poetry.core.masonry.api"
+
+[tool.poetry]
+name = "trino-python-client"
+version = "0.314.0"
+description = "Client for the Trino distributed SQL Engine"
+license = "Apache 2.0"
+authors = ["Trino Team <python-client@trino.io>"]
+readme = "README.md"
+homepage = "https://trino.io"
+repository = "https://github.com/trinodb/trino-python-client"
+
+classifiers = [
+    "Development Status :: 4 - Beta",
+    "Intended Audience :: Developers",
+    "License :: OSI Approved :: Apache Software License",
+    "Operating System :: MacOS :: MacOS X",
+    "Operating System :: Microsoft :: Windows",
+    "Operating System :: POSIX",
+    "Programming Language :: Python",
+    "Programming Language :: Python :: 3",
+    "Programming Language :: Python :: 3.7",
+    "Programming Language :: Python :: 3.8",
+    "Programming Language :: Python :: 3.9",
+    "Programming Language :: Python :: 3.10",
+    "Programming Language :: Python :: Implementation :: CPython",
+    "Programming Language :: Python :: Implementation :: PyPy",
+    "Topic :: Database :: Front-Ends",
+]
+
+packages = [
+    { include = "trino" },
+]
+
+include = ["LICENSE"]
+
+[tool.poetry.dependencies]
+keyring = { version = "*", optional = true }
+python = "^3.7"
+pytz = "*"
+requests = "*"
+requests_kerberos = { version = "*", optional = true }
+sqlalchemy = { version = "~1.4", optional = true }
```

Comment:
It was ~1.3 in setup.py previously, is there any specific reason for that change?

Reply:
@hovaesco this is _precisely_ the reason for this change, i.e., to ensure deterministic builds. Specifically when this was set to `~1.3` running, 

```
poetry run tox -e py39 -- tests/unit
```

would result in the following error:

```
Traceback:
../../../.pyenv/versions/3.9.4/lib/python3.9/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/unit/sqlalchemy/test_dialect.py:5: in <module>
    from sqlalchemy.engine import make_url
E   ImportError: cannot import name 'make_url' from 'sqlalchemy.engine'
```

In SQLAlchemy v1.4 they exposed the `make_url` function to the to the [sqlachemy.engine](https://github.com/sqlalchemy/sqlalchemy/blob/rel_1_4/lib/sqlalchemy/engine/__init__.py#L59) module as previously in v1.3 one would need to import this via, 

```
from sqlalchemy.engine.url import make_url
```

i.e., v1.3.* is not viable given how the current code is written. This logic was added in https://github.com/trinodb/trino-python-client/commit/da1441f1083ef1e908f4a8d68d2ef80491157944 which defined the `sqlalechemy~=1.3` requirement (note the inclusion of the `=` which differs from the Poetry dependency), which—per PEP 440—implies,

> For a given release identifier V.N , the compatible release clause is approximately equivalent to the pair of comparison clauses:
>
> `>= V.N, == V.*`

thus the `sqlalechemy~=1.3` requirement is equivalent to `sqlachemy>=1.3, == 1.*`. If you look at a GitHub workflow CI run ([example](https://github.com/trinodb/trino-python-client/runs/7476040790?check_suite_focus=true)) you can see, 

```
Collecting sqlalchemy~=1.3
  Downloading SQLAlchemy-1.4.39-cp39-cp39-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.6 MB)
 ```
 
that the CI is installing the latest version (v1.4.39) which adheres to the requirement specification.

Note in Poetry the [tilde requirements](https://python-poetry.org/docs/dependency-specification/#tilde-requirements)  (which don't include the `=`) differ slightly, so `~1.3` is equivalent to `>= 1.3.0, < 1.4.0`.

Reply:
I see, thanks for the detailed explanation.